### PR TITLE
[MRG+2] Save name attribute of Epochs.

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/4012

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Does this fix the mailing list issue?

**Comment by Eric89GXL** [comment]
I dug up their problem, it is like this:
```
FYI, I have my evokeds built in a conditions list loop, where the actual evokeds are stored as keys in a dictionary. 
Something like this:

for condition in conditions:
    epochs[str(condition) + '_epochs'] = mne.Epochs(params...)
    evoked[str(condition) + '_evoked']= epochs[str(condition) + '_epochs'].average()
    evoked_list.append(evoked[str(condition) + '_evoked'])

mne.write_evokeds(fname_evoked, evoked_list)
```
I'm pretty sure I was able to reproduce it.

**Comment by agramfort** [comment]
why do we need .name and .comment attributes?

**Comment by jaeilepp** [comment]
I suppose that's an usage issue. If you want to name the epochs, it should be passed as a parameter.

**Comment by jaeilepp** [comment]
As in ``mne.Epochs(params..., name=condition)``.

**Comment by agramfort** [comment]
I've never seen anyone using name=condition

maybe it was YAGNI from the start...

**Comment by Eric89GXL** [comment]
> I suppose that's an usage issue. If you want to name the epochs, it should be passed as a parameter.

I don't think so -- `epochs[key].average()` should give a reasonable `evoked.comment` regardless of (or maybe integrating) what `epochs.name` was. We just need to add to the averaging function in `BaseEpochs` that it needs to pass a reasonable `comment` to the `Evoked` constructor. It could be something like `0.1 * aud_l + 0.9 * aud_r` if there were 10% `aud_l` trials and 90% `aud_r` trials in the average, and just `aud_l` if there are only `aud_l` trials in the average.

I also doubt people were really using `epochs.name`, given that it wasn't saved. I don't feel strongly about fixing or removing it. @agramfort you favor removal?

**Comment by agramfort** [comment]
yes I would remove with proper deprecation and try to have meaning full
.comment attribute automatically

**Comment by jaeilepp** [comment]
The way I see it is that ``epochs`` is a dictionary, since Epochs does not support item assignment. If you do have an instance of Epochs with reasonable event_ids, it propagates to evoked as you suggest.

**Comment by Eric89GXL** [comment]
@jaeilepp can you clarify? I'm not sure what you're saying

**Comment by jaeilepp** [comment]
The code snippet says ``epochs[str(condition) + '_epochs'] = mne.Epochs(params...)``. In this context it looks like ``epochs`` is a dict not an instance of ``Epochs``.

If you would do ``epochs = mne.Epochs(params..., name=...)`` the name would become the comment in evoked. If you have ``dict`` as the ``event_id``, doing ``epochs['name'].average()`` the ``'name'`` will become the comment. See the tests I added.

**Comment by Eric89GXL** [comment]
Ahh, I see what you mean. I missed that in their code.

It sounds like @agramfort prefers removing `.name`. Once we do that, you'll probably need to do some logic like I outlined above to set the condition name for `Epochs(...).average()`, then.

File: mne/epochs.py [hunk 1]
```diff
@@ -829,8 +835,20 @@ def _compute_mean_or_stderr(self, picks, mode='ave'):
     def _evoked_from_epoch_data(self, data, info, picks, n_events, kind):
         """Create an evoked object from epoch data."""
         info = deepcopy(info)
+        if self.name not in ['Unknown', None]:
+            comment = self.name
+        else:
+            if len(self.event_id.keys()) == 1:
+                comment = self.event_id.keys()[0]
```
Comment (Eric89GXL) [inline mne/epochs.py]: this is a generator on Py3k so this is an error (need `list(self.event_id.keys())[0]`)

File: mne/epochs.py [hunk 2]
```diff
@@ -829,8 +835,20 @@ def _compute_mean_or_stderr(self, picks, mode='ave'):
     def _evoked_from_epoch_data(self, data, info, picks, n_events, kind):
         """Create an evoked object from epoch data."""
         info = deepcopy(info)
+        if self.name not in ['Unknown', None]:
+            comment = self.name
+        else:
+            if len(self.event_id.keys()) == 1:
```
Comment (agramfort) [inline mne/epochs.py]: if len(self.event_id) == 1:

File: mne/epochs.py [hunk 3]
```diff
@@ -829,8 +835,20 @@ def _compute_mean_or_stderr(self, picks, mode='ave'):
...
+                comment = list(self.event_id.keys())[0]
```
Comment (agramfort) [inline mne/epochs.py]: comment = next(iter(self.event_id.keys()))

**Comment by agramfort** [comment]
with the current code when I read an epo.fif file I get the deprecation warning. So for example by running the tuto plot_object_epochs.py I get the warning. We should not have deprecation warnings in the doc.

also name is deprecated to be it's weird to get:

```
In [15]: epochs['Auditory'].name
Out[15]: 'Auditory'
```

if it's deprecated is should not be used anymore.

am I missing something?

**Comment by jaeilepp** [comment]
I'm not sure what you mean. I get no warning when running ``plot_object_epochs``.

> also name is deprecated to be it's weird to get:
> 
> In [15]: epochs['Auditory'].name
> Out[15]: 'Auditory'

We should probably leave ``name`` as a private attribute at least. It is used to propagate the selection to ``evoked.comment``.

**Comment by agramfort** [comment]
I would fully deprecate the .name attribute and to propagate the .comment I
would add a .comment to Epochs so there is no naming ambiguity.

wdyt?

**Comment by jaeilepp** [comment]
I think name should still be there for 0.14 to allow standard deprecation cycle.

**Comment by agramfort** [comment]
yes ok but you should get a deprecation when trying to access the .name
attribute

**Comment by Eric89GXL** [comment]
> We should probably leave name as a private attribute at least. It is used to propagate the selection to evoked.comment.

Can you move the logic for computing this string to `.average`, or is `._name` / `._comment` going to be used other places, too?

**Comment by Eric89GXL** [comment]
It also needs to work for standard error, so really the logic should go in `_compute_mean_or_stderr`

**Comment by agramfort** [comment]
I would make name a property that uses a new comment attribute and throws a warning when you access it so we can then deprecate it safely

Clear ?

**Comment by jaeilepp** [comment]
Like this? I don't understand what you mean by comment attribute.

**Comment by agramfort** [comment]
can you check that you don't get any warning the examples or tests due to
this?

**Comment by jaeilepp** [comment]
Now the warning should only come in the tests when reading a file with name saved.

**Comment by Eric89GXL** [comment]
I'm confused, I thought the attribute was never saved (prior to this PR at
least)?

File: mne/epochs.py [hunk 4]
```diff
@@ -95,6 +95,8 @@ def _save_split(epochs, fname, part_idx, n_parts):
     last = first + len(epochs.times) - 1
     write_int(fid, FIFF.FIFF_FIRST_SAMPLE, first)
     write_int(fid, FIFF.FIFF_LAST_SAMPLE, last)
+    name = 'Unknown' if epochs._name is None else epochs._name
+    write_string(fid, FIFF.FIFF_COMMENT, name)
```
Comment (Eric89GXL) [inline mne/epochs.py]: Why add this? We're deprecating it.

**Comment by Eric89GXL** [comment]
LGTM +1 for merge

**Comment by Eric89GXL** [comment]
@agramfort feel free to merge if you're happy

**Comment by agramfort** [comment]
pushed a tiny fix  + updated what'new 

let's merge when CIs are happy

**Comment by Eric89GXL** [comment]
Fine with me, whoever sees it green first feel free to merge

**Pull request merged**